File: 3406-find-all-possible-stable-binary-arrays-i/find-all-possible-stable-binary-arrays-i.py
mod=int(1e9)+7

def solve(ones,zeros,limit,last_was_one,dp):
    if ones==0 and zeros==0:
        return 1
    
    if dp[ones][zeros][last_was_one]!=-1:
        return dp[ones][zeros][last_was_one]
    
    res=0
    if last_was_one==True:
        for i in range(1,min(zeros,limit)+1):
            res+=solve(ones,zeros-i,limit,False,dp)
    else:
        for i in range(1,min(ones,limit)+1):
            res+=solve(ones-i,zeros,limit,True,dp)
    
    dp[ones][zeros][last_was_one]=res%mod
    return res%mod
class Solution:
    def numberOfStableArrays(self, zeros: int, ones: int, limit: int) -> int:
        dp=[[ [-1 for _ in range(2)]  for _ in range(201) ]  for _ in range(201)]
        return (solve(ones,zeros,limit,True ,dp) + solve(ones,zeros,limit,False,dp))%mod 
        

File: 3406-find-all-possible-stable-binary-arrays-i/find-all-possible-stable-binary-arrays-i.cpp
class Solution {
public:
    int numberOfStableArrays(int zero, int one, int limit) {
        long long  M=1e9+7;

        long long dp[201][201][2];
        memset(dp,0,sizeof(dp));

        dp[0][0][0]=1;
        dp[0][0][1]=1;

        for (int z=0;z<=zero;z++){
            for(int o=0;o<=one;o++){

                if (z==0 && o==0)continue;

                for(int i=1;i<=min(o,limit);i++){
                    dp[z][o][0]+=dp[z][o-i][1]%M;
                }
                for(int i=1;i<=min(z,limit);i++){
                    dp[z][o][1]+=dp[z-i][o][0]%M;
                }
            }
        }

        return ( dp[zero][one][1] + dp[zero][one][0] )%M;

        
    }
};